plugins/3MFWriter/ThreeMFWorkspaceWriter.py

```diff
@@ -13,15 +15,50 @@
 from UM.i18n import i18nCatalog
 catalog = i18nCatalog("cura")
 
-from cura.Utils.Threading import call_on_qt_thread
+from .UCPDialog import UCPDialog
+from .ThreeMFWriter import ThreeMFWriter
+from .SettingsExportModel import SettingsExportModel
+from .SettingsExportGroup import SettingsExportGroup
+
+USER_SETTINGS_PATH = "Cura/user-settings.json"
 
 
 class ThreeMFWorkspaceWriter(WorkspaceWriter):
     def __init__(self):
         super().__init__()
+        self._main_thread_lock = Lock()
+        self._success = False
+        self._export_model = None
+        self._stream = None
+        self._nodes = None
+        self._mode = None
+        self._config_dialog = None
 
-    @call_on_qt_thread
-    def write(self, stream, nodes, mode=WorkspaceWriter.OutputMode.BinaryMode):
+    def _preWrite(self):
```

Comment:
Not the greatest solution, but it's fine for the mockup. When integrating this, i do think we should split off the preWrite.

Reply:
The reader classes have explicit `preRead` and `read` methods, so we should probably do the same for the writers. I didn't want to go there for the mockup, but for integration yes.

For the extra param, I don't really understand what you mean. What I think is that we should add a param to the `write` method which is the mimetype of the file to be written. Or allow the mode param (which is currently useful only for the STLs) to be any type, and then have a specific enum for the 3MF/UCP.